Behringer/BCR2000/Behringer_BCR2000.js

```diff
@@ -0,0 +1,238 @@
+var gruppo = encoderGroup1
+
+//encoders
+for(var r = 0; r < nEncoders/inarow; ++r) {
+    switch (r){
+        case 0: gruppo = encoderGroup1
+        break
+        case 1: gruppo = encoderGroup2
+        break
+        case 2: gruppo = encoderGroup3
+        break
+        case 3: gruppo = encoderGroup4
+    }
+    for (var c=0; c<inarow; ++c){        
+        var knob = surface.makeKnob(c*2, 1, 2, 2)
+        knob.setControlLayer(gruppo)
+        knob.mSurfaceValue.mMidiBinding
+        .setInputPort(midiInput).setOutputPort(midiOutput)
+        .bindToControlChange(0, 1+c+r*inarow).setTypeRelativeTwosComplement()
+        knobs.push(knob)
+
+        var button = surface.makeButton(c*2, 0, 2, 1)
+        button.setControlLayer(gruppo)
+        button.mSurfaceValue.mMidiBinding
+        .setInputPort(midiInput).setOutputPort(midiOutput)
+        .bindToControlChange (0, 33+c+inarow*r)
+        buttons.push(button)
+    }
+}
+
+//buttons
+var nbuttons = 16
+for(var rb = 0; rb < nbuttons/inarow; ++rb) {
+    for (var cb=0; cb<inarow; ++cb){
+        var button = surface.makeButton(cb*2, rb+4, 2, 1)
+        button.mSurfaceValue.mMidiBinding
+        .setInputPort(midiInput).setOutputPort(midiOutput)
+        .bindToControlChange (0, 65+cb+inarow*rb)
+        buttons.push(button)
+    }
+}
+
+//knobs
+var nknob = 24
+for(var rk = 0; rk < nknob/inarow; ++rk) {
+    for (var ck=0; ck<inarow; ++ck){
+        var knob = surface.makeKnob(ck*2, rk*3+7, 2, 2)
+        knob.mSurfaceValue.mMidiBinding
+        .setInputPort(midiInput).setOutputPort(midiOutput)
+        .bindToControlChange(0, 81+ck+rk*inarow).setTypeRelativeTwosComplement()
+        knobs.push(knob)
+    }
+}
+
+//side buttons
+var nbuttonsSide = 4
+var inArowSide = 2
+for(var rb = 0; rb < nbuttonsSide/inArowSide; ++rb) {
+    for (var cb=0; cb<inArowSide; ++cb){
+        var button = surface.makeButton(cb*2+18, rb+13, 2, 1)
+        button.mSurfaceValue.mMidiBinding
+        .setInputPort(midiInput).setOutputPort(midiOutput)
+        .bindToControlChange (0, 105+cb+inArowSide*rb)
+        buttons.push(button)
+    }
+}
+
+//footswitches
+var nFootButtons = 2
+var inArowFoot = 2
+for(var rb = 0; rb < nFootButtons/inArowFoot; ++rb) {
+    for (var cb=0; cb<inArowFoot; ++cb){
+        var button = surface.makeButton(cb*2+18, rb, 2, 1)
+        button.mSurfaceValue.mMidiBinding
+        .setInputPort(midiInput).setOutputPort(midiOutput)
+        .bindToControlChange (0, 109+cb+inArowFoot*rb)
+        buttons.push(button)
+    }
+}
+
+//-----------------------------------------------------------------------------
+// 3. HOST MAPPING - create mapping pages and host bindings
+//-----------------------------------------------------------------------------
+
+// default page
+var page = deviceDriver.mMapping.makePage('Cubase remote')
+var page_mix= page.mHostAccess.mTrackSelection.mMixerChannel
+
+//send level
+page.makeValueBinding(knobs[0].mSurfaceValue, page_mix.mSends.getByIndex(0).mLevel)

```

Comment:
Those repetitive mappings (counting from 0 to 7) would probably performing better when using a for-loop like this:
```javascript
var numStrips = 8

for(var stripIndex = 0; stripIndex < numStrips; ++stripIndex) {
	var knob = knobs[stripIndex]
	var button = buttons[stripIndex]
	var knob8 = knobs[stripIndex + 8]
	var button8 = buttons[stripIndex + 8]

	var knob16 = knobs[stripIndex + 16]
	var knob24 = knobs[stripIndex + 24]
	var button24 = buttons[stripIndex + 24]

	var sendSlot = page_mix.mSends.getByIndex(stripIndex)
	var focusQuickControl = page.mHostAccess.mFocusedQuickControls.getByIndex(stripIndex)
	var trackQuickControl = page_mix.mQuickControls.getByIndex(stripIndex)
	var cueSend = page_mix.mCueSends.getByIndex(stripIndex)

	page.makeValueBinding(knob.mSurfaceValue, sendSlot.mLevel)
	page.makeValueBinding(button.mSurfaceValue, sendSlot.mOn).setTypeToggle()
	page.makeValueBinding(knob8.mSurfaceValue, focusQuickControl)
	page.makeValueBinding(button8.mSurfaceValue, focusQuickControl).setTypeToggle()
	page.makeValueBinding(knob16.mSurfaceValue, trackQuickControl)
	page.makeValueBinding(knob24.mSurfaceValue, cueSend.mLevel)
	page.makeValueBinding(button24.mSurfaceValue, cueSend.mOn).setTypeToggle()

}
```

Reply:
ok, made and tested!